### PR TITLE
Analysis 

### PR DESCRIPTION
**Pull request by jaceybronte**
This PR cleans up and debugs the previous analysis code and provides new results. Additionally, the single layer BetaVAE and multi-layer are separated so that they can be compared.

### CODE REVIEW

File: 2.train-VAE/betavae.py [hunk 1]
```diff
@@ -14,40 +14,22 @@ class BetaVAE(nn.Module):
...
-            layers.append(nn.Linear(in_dim, hidden_dim))
-            layers.append(nn.BatchNorm1d(hidden_dim))
-            layers.append(nn.ReLU())
-            in_dim = hidden_dim
-        # The final layer outputs the reconstructed input
-        layers.append(nn.Linear(in_dim, input_dim))
-        layers.append(nn.Sigmoid())
-        self.decoder = nn.Sequential(*layers)
+        self.encoder = nn.Sequential(
+            nn.Linear(input_dim, latent_dim * 2),
```
Comment (MikeLippincott) [inline 2.train-VAE/betavae.py:24]: Why multiply by two???

File: 2.train-VAE/betavae_multilayer.py [hunk 2]
```diff
@@ -0,0 +1,267 @@
...
+
+    Args:
+        input_dim (int): Dimension of the input data.
+        latent_dim (int): Dimension of the latent space.
+        beta (float): Weight for the Kullback-Leibler divergence term in the loss function.
+        hidden_dim (int): Dimension of the hidden layer. (utilized for multiple layers)
+        num_layers (int): Number of layers
+    """
+
+    def __init__(self, input_dim, latent_dim, beta, hidden_dim, num_layers):
```
Comment (MikeLippincott) [inline 2.train-VAE/betavae_multilayer.py]: Consider adding type hints to these inputs

Comment (MikeLippincott) [inline 2.train-VAE/betavae_multilayer.py]:
```suggestion
    def __init__(self, input_dim: int, latent_dim: int, beta: float, hidden_dim: int, num_layers: int):
```

File: 2.train-VAE/betavae_multilayer.py [hunk 3]
```diff
@@ -0,0 +1,267 @@
...
+        self.beta = beta
+
+        # Create encoder with variable number of layers
+        layers = []
+        in_dim = input_dim
+        for _ in range(num_layers):
+            layers.append(nn.Linear(in_dim, hidden_dim))
+            layers.append(nn.BatchNorm1d(hidden_dim))
+            layers.append(nn.ReLU())
+            in_dim = hidden_dim
```
Comment (MikeLippincott) [inline 2.train-VAE/betavae_multilayer.py:35]: The hidden layers all have the same amount of nodes. Would it be worth while to change this to a variable amount of nodes??

File: 2.train-VAE/betavae_multilayer.py [hunk 4]
```diff
@@ -0,0 +1,267 @@
...
+        # The final layer outputs the reconstructed input
+        layers.append(nn.Linear(in_dim, input_dim))
+        layers.append(nn.Sigmoid())
+        self.decoder = nn.Sequential(*layers)
+        self.latent_dim = latent_dim
+        self.beta = beta
+
+    def reparameterize(self, mu, log_var):
```
Comment (MikeLippincott) [inline 2.train-VAE/betavae_multilayer.py]: consider type hints here

File: 2.train-VAE/betavae_multilayer.py [hunk 5]
```diff
@@ -0,0 +1,267 @@
...
+        test_loader (DataLoader): DataLoader for the test data.
+
+    Returns:
+        Average loss over the test dataset.
+    """
+    model.eval()
+    val_loss = 0
+    with torch.no_grad():
+        for batch in val_loader:
+            data = batch[0]
```
Comment (MikeLippincott) [inline 2.train-VAE/betavae_multilayer.py:170]: why grab the first element?? is this a single dimension or multiple??? is there a label or non feature space in the second element of the batch?

File: 2.train-VAE/betavae_multilayer.py [hunk 6]
```diff
@@ -0,0 +1,267 @@
...
+        for batch in data_loader:
+            data = batch[0]
+            mu, _ = model.encode(data)
+            latent_space.append(mu.cpu().numpy())
+
+    latent_space = np.concatenate(latent_space, axis=0)
+    latent_df = pd.DataFrame(latent_space)
+    latent_df.insert(0, 'ModelID', metadata['ModelID'])
+
+    latent_df_dir = pathlib.Path("./results/latent_df_multi.parquet")
```
Comment (MikeLippincott) [inline 2.train-VAE/betavae_multilayer.py:223]: consider passing this path in as an arg. Hardcoded paths are less flexible. What if you want to move where you write the data to in the future?

Comment (MikeLippincott) [reply 2.train-VAE/betavae_multilayer.py:223]: Also consider resolving the path and checking if the parent folder is present prior to writing to avoid errors

File: 2.train-VAE/betavae_multilayer.py [hunk 7]
```diff
@@ -0,0 +1,267 @@
...
+        subset_train_df: the qc training dataframe
+
+    Returns:
+        Gene weight dataframe
+    """
+    weight_matrix = model.encoder[0].weight.detach().cpu().numpy().T  # Transpose the weight matrix
+    weight_df = pd.DataFrame(weight_matrix)
+
+    # Save as parquet to use for heatmap
+    weight_df_dir = pathlib.Path("./results/weight_matrix_encoder.parquet")
```
Comment (MikeLippincott) [inline 2.train-VAE/betavae_multilayer.py:244]: same here

File: 3.analysis/results/combined_gsea_results.parquet.gz
Comment (MikeLippincott) [inline 3.analysis/results/combined_gsea_results.parquet.gz:1]: Consider adding only crucial data to github to avoid repo bloating. Best to talk with @gwaybio for best strategies for which files to push.

File: utils/data_loader.py [hunk 8]
```diff
@@ -92,36 +93,37 @@ def load_train_test_data(
...
+        test_file = test_file.filter(gene_list_passed_qc, axis=1)
+        val_file = val_file.filter(gene_list_passed_qc, axis=1)
 
     # Normalize data
     
     if zero_one_normalize == True:
-        train_data = train_data.values.astype(np.float32)
-        test_data = test_data.values.astype(np.float32)
-        val_data = val_data.values.astype(np.float32)
+        train_file = train_file.values.astype(np.float32)
```
Comment (MikeLippincott) [inline utils/data_loader.py:118]: Why change from data to file? Semanticly it is not a file?

File: 1.data-exploration/figures/controlgsea.png
Comment (MikeLippincott) [inline 1.data-exploration/figures/controlgsea.png:1]: Nice! It would be cool to see x and y axis cut off for LFC and adj. pval significance!

File: 1.data-exploration/figures/heatmap.png
Comment (MikeLippincott) [inline 1.data-exploration/figures/heatmap.png:1]: It is a bit hard to read the side bar labels and some colours are hard to see IMO.

**Review by MikeLippincott — APPROVED** [review]
LGTM, ready to merge after reviewing comments. I ask more questions than give comments. Once you can answer the questions it should be clear if changes are needed or not.

File: 2.train-VAE/betavae.py (hunk 1 above)
Comment (jaceybronte) [reply 2.train-VAE/betavae.py:24]: The * 2 should be for outputting two separate vectors of size latent_dim that are then split with h.chunk(2, dim=-1) so that mu has the first and log_var has the second.

File: 2.train-VAE/betavae_multilayer.py (hunk 2 above)
Comment (jaceybronte) [reply 2.train-VAE/betavae_multilayer.py]: Corrected

Comment (jaceybronte) [reply 2.train-VAE/betavae_multilayer.py]: Corrected

File: 2.train-VAE/betavae_multilayer.py (hunk 3 above)
Comment (jaceybronte) [reply 2.train-VAE/betavae_multilayer.py:35]: Might be! will try it out

File: utils/data_loader.py (hunk 8 above)
Comment (jaceybronte) [reply utils/data_loader.py:118]: Allows it to return original file if there is no desire to normalize or drop columns
